### PR TITLE
Authentication and auditing

### LINKED ISSUE #157: Support authenticated users and auth trail
- [x] Need to add the "created by" and "updated by fields"
- [x] Pulling user email out of Okta token and saving to user table
- [x] Ensure user is authenticated before any action can be taken

### PR DESCRIPTION
**Pull request by benwarfield-usds**
Added audit fields for created/updated by to all managed entities, and updated `getCurrentUser` to fetch the actual current user.

TODO:
- [x] all the tests are broken
- [x] test with JWT
- [x] make JWT work
- ~this might be where we fix `getCurrentOrganization` so it doesn't keep returning dummy data after our prod launch.~ => #170

Resolves #157.

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuditingConfig.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
+
+import gov.cdc.usds.simplereport.db.model.ApiUser;
+import gov.cdc.usds.simplereport.service.ApiUserService;
+
+@Configuration
+@EnableJpaAuditing
+public class AuditingConfig {
+
+	public static final String ID_PROVIDER = "currentUserProvider";
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/AuditingConfig.java:20]: We don't actually use this, as it turns out—probably should be axed.

File: backend/src/main/java/gov/cdc/usds/simplereport/api/organization/OrganizationResolver.java [hunk 2]
```diff
@@ -31,9 +33,6 @@ public Organization getOrganization() {
 
         List<DeviceType> currentDevices = _deviceService.fetchDeviceTypes();
         List<String> currentDeviceNames = currentDevices.stream().map(d->d.getName()).collect(Collectors.toList());
-        for (int i = 0; i < currentDevices.size(); i++) {
-            currentDeviceNames.add(currentDevices.get(i).getName());
-        }
         if (!currentDeviceNames.contains("Abbott IDNow")) {
             _deviceService.createDeviceType("Abbott IDNow", "ID Now", "Abbott");
```
Comment (timothybest-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/organization/OrganizationResolver.java:37]: can this be called when the device already exists?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiUser.java [hunk 3]
```diff
@@ -1,53 +1,75 @@
...
+import org.hibernate.annotations.UpdateTimestamp;
+
+import gov.cdc.usds.simplereport.db.model.auxiliary.PersonName;
 
 /**
  * The bare minimum required to link an authenticated identity to actions and data
  * elsewhere in the schema.
+ * 
+ * Since we can't use JPA auditing here, we use Hibernate {@link CreationTimestamp} and
+ * {@link UpdateTimestamp}, and feel kind of weird about it.
```
Comment (nickrobison-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiUser.java:26]: Interested in this comment, but not worth blocking a PR.

**Review by timothybest-usds — APPROVED** [review]

**Review by nickrobison-usds — APPROVED** [review]
This looks good to me! Your JPA annotation-fu is far beyond mine!

I think this looks like a simple and solid approach. Nice work!

File: backend/src/main/java/gov/cdc/usds/simplereport/api/organization/OrganizationResolver.java (hunk 2 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/organization/OrganizationResolver.java:37]: Not safely.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiUser.java (hunk 3 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiUser.java:26]: It is possible there's a way to do this safely and I didn't find it last night: basically, the JPA audit assumes that you can get all the audit information up front and then apply it as needed. The `ApiUser` object is part of the audit info for JPA; if you are in the process of creating an `ApiUser` and then JPA says "oh hey, I need to get the current `ApiUser` in order to finish this, and then it starts to create an `ApiUser`... you get a stack overflow.

**Pull request merged**